corehq/messaging/scheduling/models/content.py

```diff
@@ -133,9 +137,10 @@ def send(self, recipient, logged_event, phone_entry=None):
             logged_subevent.error(MessagingEvent.ERROR_NO_MESSAGE)
             return
 
-        email_address = recipient.get_email()
-        if not email_address:
-            logged_subevent.error(MessagingEvent.ERROR_NO_EMAIL_ADDRESS)
+        try:
+            email_address = self.get_recipient_email(recipient)
+        except ContentException as e:
+            logged_subevent.error(e.error_type, additional_error_text=e.additional_text)
```

Comment:
Just thinking if it was worth adding a new exception specifically for this case here. It nice that this just adds one more line to this long function but I am not sure if its really helping. Just laying down the 2 approaches I could think of for comparison, with no intention that the approach taken should be changed.

Approach 1
```
email_address = self._get_recipient_email_address(recipient, logged_subevent)
if not email_address:
    return
```

```
def _get_recipient_email_address(recipient, logged_subevent):
    email_address = recipient.get_email()
    if not email_address:
        logged_subevent.error(MessagingEvent.ERROR_NO_EMAIL_ADDRESS)
    else:
        try:
            validate_email(email_address)
        except ValidationError as exc:
            logged_subevent.error(MessagingEvent.ERROR_INVALID_EMAIL_ADDRESS, str(exc))
            email_address = None
    return email_address
```

Approach 2 (to keep `logged_subevent` to log error in this function only):
```
email_address = recipient.get_email()
if not email_address:
    logged_subevent.error(MessagingEvent.ERROR_NO_EMAIL_ADDRESS)
    return

try:
    validate_email(email_address)
except ValidationError as exc:
    logged_subevent.error(MessagingEvent.ERROR_INVALID_EMAIL_ADDRESS, str(exc))
    return
```

Reply:
A big challenge with both of those approaches is testing. One of the main reasons I put this in a separate function that doesn't update the subevent directly is that it makes it significantly easier to test (no subevent needs to be created). I initially did just have the code inline in the main function but testing that would require a large amount of setup of DB models.

Reply:
Cool, thanks @snopoke.